File: myAtoi.java
public class Solution {
	private static final int upperBound = Integer.MAX_VALUE / 10;
    public int myAtoi(String str) {
    	str = str.trim();
    	int i = 0, n = str.length();
    	int sign = 1;
    	if (i < n && str.charAt(i) == '+') {
    		i++;
    	} else if (i < n && str.charAt(i) == '-') {
    		i++;
    		sign = -1;
    	}
    	int num = 0;
    	while (i < n && Character.isDigit(str.charAt(i))) {
    		int digit = Character.getNumericValue(str.charAt(i));
    		if (num > upperBound || num == upperBound && digit >= 8) {
				return sign == 1 ? Integer.MAX_VALUE : Integer.MIN_VALUE;    			
    		}
    		num = num * 10 + digit;
    		i++;
    	}
    	return sign * num;
    }
}

File: myAtoi.py
class Solution(object):
    def myAtoi(self, str):
        """
        :type str: str
        :rtype: int
        """
        MAX_VAL, MIN_VAL = 2147483647, -2147483648 
        str = str.strip()
        i,n = 0, len(str)

        sign = 1
        if i < n and (str[i] == '+'):
        	i += 1
        elif i< n and (str[i] == '-'):
        	i += 1
        	sign = -1

        num = 0
        upperbound = MAX_VAL // 10
        while i < n and str[i].isdigit():
        	digit = int(str[i])
        	if num > upperbound or (num == upperbound and digit >= 8):
        		return MAX_VAL if sign == 1 else MIN_VAL
        	num = num * 10 + digit
        	i += 1
        return num * sign